### ISSUE TITLE
RFC: Environment-owned reward

### ISSUE THREAD

**Issue by qgallouedec** [huggingface/trl#5912]
# RFC: Environment-owned reward

**Scope:** `GRPOTrainer` + `AsyncGRPOTrainer`. Same pattern applies to any online trainers with environment support.
**Out of scope:** Environment-owned dataset (RFC #5903)

## Summary

Let the environment own the reward.
Today the trainer owns reward computation: you pass `reward_funcs` to the trainer, and the trainer scores each completion. This RFC lets the environment return the reward instead.

In the RL-native formulation, where the environment that runs the episode is also the one that scores it.

This matters for multi-turn / stateful environments: the reward is usually a function of the environment's internal state (did the game end in a win? was the word guessed?), which the trainer does not have. Forcing `reward_funcs` then means leaking env state back out just so an external function can recompute what the env already knows.

## Other data points

- verifiers: the environment owns the reward, via a `rubric` (`score_rollout` / `score_group`).
- OpenEnv: gym-like — `step()` returns the reward directly. The environment owns it by design.

### Before

<img width="1729" height="1040" alt="Image" src="https://github.com/user-attachments/assets/cdddde54-b0a0-4d87-a305-934f24a6777d" />

### After

<img width="1728" height="1040" alt="Image" src="https://github.com/user-attachments/assets/655eacd9-bac4-430a-bb28-20f42f3a6370" />

**Before:** the rollout's completions are sent back to the trainer-owned `reward_funcs`, along with the environment itself, which score them (and often only need the env state to do so).
**After:** the environment, which already holds the state, returns the reward directly. The trainer still turns rewards into advantages that part is unchanged.

### Example: wordle-like game

The reward (was the word guessed?) can be written as a function, but it needs the target word, which the environment already knows.

#### Before

The reward function lives in the trainer and must receive the target through an access to the environment (`environments.target`)

```python
def solved_reward(completions, environments, **kwargs):
    return [1.0 if env.target in completion else 0.0 for completion, env in zip(completions, environments)]

trainer = GRPOTrainer(
    model=my_model,
    reward_funcs=solved_reward,              # trainer owns the reward
    train_dataset=dataset, 
    environment_factory=WordleEnvironment,
)
```

#### After

The environment scores the episode it just ran:

```python
class WordleEnvironment:
    def reset(self) -> None:
        self._target = sample(list_of_words)
        self._solved = False

    def guess(self, word: str) -> str:
        """Submit a guess, get per-letter feedback."""
        self._solved = word == self._target
        ...

    def get_reward(self) -> float:               # env knows the target → scores itself
        return 1.0 if self._solved else 0.0

trainer = GRPOTrainer(
    model=my_model,
    environment_factory=WordleEnvironment,   # owns the reward, reward_funcs becomes optional
)
```

### We could still combine both

This new design would allow both to coexist: the environment can return a reward, and the trainer can still have `reward_funcs`:

```python
def format_reward(completions, **kwargs):
    return [bool(check_style(completion)) for completion in completions]

trainer = GRPOTrainer(
    model=my_model,
    reward_funcs=format_reward,              # trainer-owned reward (e.g. style check)
    environment_factory=WordleEnvironment,   # env-owned reward (e.g. solved or not)
)
```

in the above axample you'd have two reward sources: the environment's via `get_reward()` and the trainer's `format_reward()`

## Alternative design: rubrics-based

Instead of a single `get_reward()` returning one scalar, the environment could own **several** reward components, a "rubric", as in verifiers. For example:

```python
class WordleEnvironment:
    def reset(self) -> None: ...
    def guess(self, word: str) -> str: ...

    # multiple reward components, all owned by the env
    def reward_solved(self) -> float:
        return 1.0 if self._solved else 0.0

    def reward_efficiency(self) -> float:
        return 1.0 - self._n_guesses / 6
```

This raises a few questions:
- how to "register" these components so the trainer can find them? Naming convention (`reward_` prefix)? Class-level registry (e.g. `WordleEnvironment.rewards = ["solved", "efficiency"]`)?
- ...

Personally, I'm not super aligned on this one. I feel like would sacrifice a lot of simplicity for something that we can do with the first design (single `get_reward()`) + some internal env logic. The trainer doesn't need to know about multiple components if the env can aggregate them into a single reward signal internally.

## What this would imply

- A reserved `get_reward()` method on the environment (parallel to `reset()`), called once per completed rollout. When present, it is a reward source like any `reward_func`.
- The trainer's job is unchanged downstream: it still turns rewards into advantages (group normalization, weighting, logging).

## Open questions so far

1. `get_reward()` signature — does it receive the completion (`get_reward(completion=...)`), or rely purely on internal env state? (Recommend: only internal state)
2. Coexistence and weights: if both `get_reward()` and `reward_funcs` are present, how do we combine them? How does it play with reward weighting (`reward_weights`)?
3. Rubrics-based (multiple env-owned components): who aggregates the components into the final reward — the env or the trainer? (See "Alternative design: rubrics-based" — this drives whether the env returns a scalar or named components, and where `reward_weights` / per-component logging live.)

**Comment by qgallouedec** [comment]
closed by https://github.com/huggingface/trl/pull/6238
